File: Leetcode/Algorithms/Easy/MaximumProductDifferenceBetweeTwoPairs/MaximumProductDifferenceBetweenTwoPairs.cpp
class Solution {
public:
    int maxProductDifference(vector<int>& nums) {
        int max1 = INT_MIN; 
        int max2 = INT_MIN;
        int min1 = INT_MAX; 
        int min2 = INT_MAX;
        for(int &element: nums)   
        {
            if(element > max1)
            {
                max2 = max1;
                max1 = element;
            }
            else if(element > max2)
            {
                max2 = element;
            }
            
            if(element < min1)
            {
                min2 = min1;
                min1 = element;
            }
            else if(element < min2)
            {
                min2 = element;
            }
        }
        return (max1 * max2) - (min1 * min2);
    }
};

// Time Complexity - O(N)
// Space Complexity - O(1)

File: LeetCode/Algorithms/Easy/MaximumProductDifferenceBetweeTwoPairs/MaximumProductDifferenceBetweenTwoPairs.py
class Solution(object):
    def maxProductDifference(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums.sort()
        n = len(nums)
        a, b, c, d = nums[n-1], nums[n-2], nums[0], nums[1]
        
        return (a*b) - (c*d)
        
# Time Complexity - O(nlogn)
# Space Complexity - O(1)

